### ISSUE TITLE
Core Engine

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#621]
See spec at https://github.com/nunit/dev/wiki/Core-Engine

**Comment by oznetmaster** [comment]
So the core engine will support multiple/separate AppDomain testing, but not multiple/separate process?

**Comment by CharliePoole** [comment]
Not multiple but possibly separate. The table is my best guess right now.

Any runner that can run tests in an arbitrary directory needs to support creating an AppDomain, at least so long as we stick with using Load rather than LoadFrom to load the test. It appears that the Xamarin runners work similarly to NUnitLite and don't need a separate domain.

The key is what the VS adapter needs. Since we don't control the AppDomain in which the adapter is executed, we have to create one. If the engine doesn't do it, then the adapter would have to do it and then launch the engine in that domain.

I'm most comfortable seeing how it works out in the code rather than saying for sure right now.

**Comment by oznetmaster** [comment]
I thought we currently used both Load and LoadFrom, determined by the provided assembly name?  Or is that only in NUnitLiteRunner?

 

Do you see the core engine being used in the NUnitLiteRunner to refactor its functionality?

**Comment by CharliePoole** [comment]
Historically, we never used LoadFrom to load test assemblies. In the current code we are using LoadFrom in NETCF only. In other builds we get the AssemblyName from the path and then Load.

The reasons we have avoided LoadFrom relate to problems arising when an assembly in a LoadFrom context Loads another assembly, passing it a reference back to one of it's own classes. We fixed the problem by not using LoadFrom. It's not the only fix, but it's definitely the simplest.
